signer/src/transaction_coordinator.rs

```diff
@@ -73,4 +85,189 @@
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    /// Construct and coordinate WSTS signing rounds for sBTC transactions on Bitcoin,
+    /// fulfilling pending deposit and withdraw requests.
+    #[tracing::instrument(skip(self))]
+    async fn construct_and_sign_bitcoin_sbtc_transactions(
+        &mut self,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+    ) -> Result<(), error::Error> {
+        let fee_rate = self.bitcoin_client.estimate_fee_rate().await?;
+
+        let aggregate_key = self.get_aggregate_key(bitcoin_chain_tip).await?;
+
+        let signer_btc_state = self.get_btc_state(fee_rate, &aggregate_key).await?;
+
+        let pending_requests = self.get_pending_requests(signer_btc_state).await?;
+
+        let transaction_package = pending_requests.construct_transactions()?;
+
+        for transaction in transaction_package {
+            self.sign_and_broadcast(aggregate_key, transaction).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Construct and coordinate signing rounds for
+    /// `deposit-accept`, `withdraw-accept` and `withdraw-reject` transactions.
+    #[tracing::instrument(skip(self))]
+    async fn construct_and_sign_stacks_sbtc_response_transactions(
+        &mut self,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+    ) -> Result<(), error::Error> {
+        // TODO(320): Implement
+        todo!();
+    }
+
+    /// Coordinate a signing round for the given request
+    /// and broadcast it once it's signed.
+    #[tracing::instrument(skip(self))]
+    async fn sign_and_broadcast(
+        &mut self,
+        aggregate_key: p256k1::point::Point,
+        transaction: utxo::UnsignedTransaction<'_>,
+    ) -> Result<(), error::Error> {
+        let _coordinator_state_machine = wsts_state_machine::CoordinatorStateMachine::load(
+            &mut self.storage,
+            aggregate_key,
+            self.signer_public_keys.clone(),
+            self.threshold,
+            self.private_key,
+        )
+        .await?;
+        // TODO(319): Coordinate signing round and broadcast result
+        todo!();
+    }
+
+    // Determine if the current coordinator is the coordinator
+    //
+    // Note that this function is technically not fallible,
+    // but for now we have chosen to return phantom errors
+    // instead of adding expects/unwraps in the code.
+    // Ideally the code should be formulated in a way to guarantee
+    // it being infallible without relying on sequentially coupling
+    // expressions. However, that is left for future work.
+    fn is_coordinator(
+        &self,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+    ) -> Result<bool, error::Error> {
+        let mut hasher = sha2::Sha256::new();
+        hasher.update(bitcoin_chain_tip);
+        let digest = hasher.finalize();
+        let index =
+            usize::from_be_bytes(*digest.first_chunk().ok_or(error::Error::TypeConversion)?);
```

Comment:
Nice.

The `core::slice::first_chunk` function is generic, and returns an option of an array of fixed length where the length is the generic parameter (here the length is inferred to be 8 on a typical machine). `None` is only returned if the generic length parameter is longer than the length of the underlying slice. Since the Sha256 type contains 32 bytes, this should never error (which you noted already).

Also, using the bitcoin block hash directly is no good since it isn't random enough (although it's 32 bytes, it typically starts with lots of leading zeros bits). So we hash it first and use the first 8 bytes of that to make sure that it actually is random. I would add a comment somewhere noting this.

Reply:
Right, actually I think the block hash in itself is probably good enough. Definitely for v1 :sweat_smile: Though there's no harm in the extra hashing of it. I'll update the comment.

Reply:
Actually, I'll skip the hashing. I don't see any reason for doing it.

Reply:
Changed my mind (or rather you did in a DM in slack), we'll keep the hashing. I've added an explanatory comment now.